### ISSUE TITLE
DistributedDataParallel

### ISSUE THREAD

**Issue by LanceGe** [HobbitLong/SupContrast#26]
Is there a way to make SupContrast work for DistributedDataParallel? By default each worker can only see its own sub-batch so the inter-sub-batch relationship of the samples will be utilized.

**Comment by HobbitLong** [comment]
You can use `all_gather` to gather features together. The caveat is that you need to manually propagate gradients through `all_gather` op, as it doesn't auto-bp.

**Comment by LanceGe** [comment]
> You can use `all_gather` to gather features together. The caveat is that you need to manually propagate gradients through `all_gather` op, as it doesn't auto-bp.

I finally make it work with the help of [diffdist](https://github.com/ag14774/diffdist), which provides a differentiable all_gather wrapper.

**Comment by ShijianXu** [comment]
Hi, can you share your code about how to implement this? I am not familiar with `all_gather` .etc operations. Thanks a lot.

**Comment by LanceGe** [comment]
> Hi, can you share your code about how to implement this? I am not familiar with `all_gather` .etc operations. Thanks a lot.

First, install [diffdist](https://github.com/ag14774/diffdist). 
Then put the following snippet before calling the criterion: 
`
        import diffdist.functional as distops

        features = distops.all_gather(
            gather_list=[torch.zeros_like(features) for _ in range(torch.distributed.get_world_size())],
            tensor=features,
            next_backprop=None,
            inplace=True,
        )
        features = torch.cat(features)

        labels = distops.all_gather(
            gather_list=[torch.zeros_like(labels) for _ in range(torch.distributed.get_world_size())],
            tensor=labels,
            next_backprop=None,
            inplace=True,
        )
        labels = torch.cat(labels)
`

**Comment by ShijianXu** [comment]
Thank you for your quick reply.

So, then I can simply compute the loss as usual and then backward the gradient?

**Comment by LanceGe** [comment]
Yes, but I'm not sure this is bug-free.

**Comment by ShijianXu** [comment]
OK. Anyway, thanks a lot.

**Comment by ShijianXu** [comment]
Just for reference, [this](https://github.com/open-mmlab/OpenSelfSup/blob/master/openselfsup/models/utils/gather_layer.py) seems to be a reliable solution.
